### PR TITLE
Real-time bidding in double loop

### PR DESCRIPTION
**Pull request by xiangao1**
## Summary/Motivation:
Solve real-time bidding problems and bid into the real-time market in the double loop.

We should merge #824 before this PR.

## Changes proposed in this PR:
- Enable real-time bidding in the double loop
- When there is not enough dispatch/price signal from the market (Prescient), deactivate/unfix the corresponding tracking and bidding constraints/variables.
- Propose template (abstract) class for price forecaster.
- Update examples, docs, and tests.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by ksbeattie** [comment]
@xiangao1 I've put this on the IDAES August release board assuming it is NOT expected for the IDAES May release meaning it will not be in the DISPATCHE June release.  Let me know if that is not correct!

**Comment by xiangao1** [comment]
> @xiangao1 I've put this on the IDAES August release board assuming it is NOT expected for the IDAES May release meaning it will not be in the DISPATCHE June release. Let me know if that is not correct!

Thanks, @ksbeattie ! I anticipate this draft PR still needs some revisions. @adowling2 may have more to share about the plan.

**Comment by ksbeattie** [comment]
@xiangao1, will this be ready for the Aug release?  As in, within the next week?

**Comment by xiangao1** [comment]
@ksbeattie I reviewed the conflicts listed here. Some of the conflicts are changes in the doc strings and import and some of them are introduced due to new capabilities. How should I address these conflicts?

**Comment by ksbeattie** [comment]
> @ksbeattie I reviewed the conflicts listed here. Some of the conflicts are changes in the doc strings and import and some of them are introduced due to new capabilities. How should I address these conflicts?

The short answer is that they should be resolved so that both your changes and the changes from main are preserved.

But first a question: Are all the conflicts you are seeing here involving changes you've made in this branch?

I ask because this is a branch that is coming from a fork (yours) of a fork (Robby's) which can become rather confusing.  It would be better for you to have your changes coming from a branch of your own direct fork from the primary idaes-pse upstream repo.

**Comment by xiangao1** [comment]
@ksbeattie 
Are all the conflicts you are seeing here involving changes you've made in this branch?
Yes, all the conflicts are from this branch. On my local machine, my remotes are my fork and the idaes fork. I am not tracking Robby's fork anymore.

**Comment by xiangao1** [comment]
Hi @adowling2! Could you review this? I just formatted the code with Black.

**Comment by ksbeattie** [comment]
@xiangao1, @adam-a-a  We're planning on creating the IDAES Aug release candidate this week and this doesn't look ready.

**Comment by adowling2** [comment]
I should have a review done by the end of today.

File: idaes/apps/grid_integration/coordinator.py
Comment (adowling2) [inline idaes/apps/grid_integration/coordinator.py:476]: Remove this commented out line?

**Review by adowling2 — APPROVED** [review]

**Comment by adowling2** [comment]
@radhakrishnatg @jghouse88 @jsiirola Any chance we could get a quick review?

**Comment by adowling2** [comment]
@ksbeattie we just need another reviewer

**Review by dguittet — APPROVED** [review]
FWIW, I've updated Radhakrishna's [nuclear flowsheet](https://github.com/gmlc-dispatches/dispatches/blob/main/dispatches/case_studies/nuclear_case/nuclear_flowsheet_double_loop.ipynb) to use this branch + #870 to check that it still runs. Not a lot of changes were required: a few minor interface changes, and using "Backcaster", which requires historical prices. 

I've attached the Git Patch file that contains the changes I made + some comments:
[upgrade_nuclear_double_loop.patch.zip](https://github.com/IDAES/idaes-pse/files/9378909/upgrade_nuclear_double_loop.patch.zip) 
 in case someone wants to apply these changes themselves and see what it looks like, if it runs, etc.

**Comment by radhakrishnatg** [comment]
Thank you @dguittet for reviewing the PR and for testing it with the nuclear case study! After this PR is merged, would you be able to open a PR in DISPATCHES with these changes (You will have to update your fork/branch to reflect the repo reorganization)?

@xiangao1 Looks like some tests are failing in the bidder code. Could you please take a look at it?

**Comment by lbianchi-lbl** [comment]
@radhakrishnatg @xiangao1 as far as I can understand the failing test is due to `tx_utils` (used for `validate_and_clean_cost_curve()` in `bidder.Bidder._assemble_bids()`) being imported conditionally only if `egret` is installed:

https://github.com/xiangao1/idaes-pse/blob/a52b074ab64f15aba073a657576ddd2856de75d0/idaes/apps/grid_integration/bidder.py#L1201-L1216

https://github.com/xiangao1/idaes-pse/blob/a52b074ab64f15aba073a657576ddd2856de75d0/idaes/apps/grid_integration/bidder.py#L22-L24

`egret` will be available in the developer installation (where most if not all of IDAES's optional dependencies are available), but not in the environment where the check "Run tests from site-packages (non-editable) installation" (which is designed for this exact purpose) is run.

Apart from skipping the tests that (directly or indirectly) result in `Bidder._assemble_bids()` being called with `pytest.importorskip()`, I would also consider adding a better error message (i.e. something more meaningful than the current `NameError: name "tx_utils" is not defined`) inside `Bidder._assemble_bids()` before `tx_utils.validate_and_clean_cost_curve()` is called, e.g.:

```py
if not egret_available:
    raise RuntimeError("'egret' must be installed to use this functionality")
```

**Pull request merged**
